SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1020 @@
+  private func globalInitMayConflictWith(
+    sideEffect: Instruction,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    switch sideEffect {
+    case let storeInst as StoreInst:
+      return mayReadOrWrite(address: storeInst.destinationOperand.value, aliasAnalysis)
+    case let loadInst as LoadInst:
+      return mayWrite(toAddress: loadInst.operand.value, aliasAnalysis)
+    case is CondFailInst:
+      return false
+    default:
+      return true
+    }
+  }
+  
+  /// Returns `true` if any of the instructions in `sideEffects` cannot be
+  /// reordered with a call to this global initializer (which is in the same basic
+  /// block).
+  func globalInitMayConflictWith(
+    blockSideEffectSegment: StackWithCount<Instruction>.Segment,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    return blockSideEffectSegment
+      .contains { sideEffect in
+        globalInitMayConflictWith(
+          sideEffect: sideEffect,
+          aliasAnalysis
+        )
+      }
+  }
+
+  /// Returns `true` if any of the instructions in `loopSideEffects` which are
+  /// post-dominated by a call to this global initializer cannot be reordered with
+  /// the call.
+  func globalInitMayConflictWith(
+    loopSideEffects: StackWithCount<Instruction>,
+    notPostDominatingPreheader: BasicBlock,
+    _ aliasAnalysis: AliasAnalysis,
+    _ postDomTree: PostDominatorTree
+  ) -> Bool {
+    guard parentBlock.postDominates(notPostDominatingPreheader, postDomTree) else {
+      return true
+    }
+
+    return loopSideEffects
+      .contains { sideEffect in
+        // Only check instructions in blocks which are "before" (i.e. post-dominated
+        // by) the block which contains the init-call.
+        // Instructions which are before the call in the same block have already
+        // been checked.
+        parentBlock.strictlyPostDominates(
+          sideEffect.parentBlock,
+          postDomTree
+        ) && globalInitMayConflictWith(
+          sideEffect: sideEffect,
+          aliasAnalysis
+        )
+      }
+  }
+  
+  func isInAccessScope(of beginAccess: BeginAccessInst, _ domTree: DominatorTree) -> Bool {
+    return beginAccess.dominates(self, domTree) && beginAccess.endAccessInstructions
+      .allSatisfy { endAccessInst in
+        self.dominates(endAccessInst, domTree)
+      }
+  }
+}
+
+private extension UnaryInstruction {
+  /// Returns true if `sideEffects` contains any memory writes which
+  /// may alias with the memory addressed by this instruction.
+  func mayWriteTo(
+    sideEffects: StackWithCount<Instruction>,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    return sideEffects
+      .contains { sideEffect in
+        sideEffect.mayWrite(toAddress: operand.value, aliasAnalysis)
+      }
+  }
+}
+
+private extension StoreInst {
+  /// Returns a `true` if this store is a store to `accessPath`.
+  func storesTo(_ accessPath: AccessPath) -> Bool {
+    guard self.storeOwnership != .initialize else {
```

Comment:
This check doesn't belong here. Except you rename the function to `isInitializingStoreTo`. Also, shouldn't trivial and none ownership stores also be allowed?